File: app/src/main/java/com/example/firebasestudy/InscriptionActivity.java
package com.example.firebasestudy;

import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.text.TextUtils;
import android.util.Log;
import android.util.Patterns;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.example.firebasestudy.databinding.ActivityInscriptionBinding;
import com.example.firebasestudy.model.User;
import com.google.android.gms.tasks.Continuation;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.android.material.snackbar.Snackbar;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;

import java.io.IOException;

public class InscriptionActivity extends AppCompatActivity {
    private static final int PICK_IMAGE_REQUEST = 1000 ;
    private ActivityInscriptionBinding binding ;
    private Uri filePath;
    private FirebaseAuth mAuth;
    private FirebaseDatabase mDatabase;
    private DatabaseReference mRefrence;
    private FirebaseStorage mStorage;
    private User mUser = new User();


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityInscriptionBinding.inflate(getLayoutInflater());
        View view = binding.getRoot();
        setContentView(view);
        initFireBAse();
        binding.addimage.setOnClickListener(this::choosePicture);
        binding.inscriBtn.setOnClickListener(v -> save());
    }

    private void save (){
        if (!isEmpty()){
            String mail = binding.mailInscri.getText().toString();
            String pwd = binding.pwdInscri.getText().toString();

            if (validEmail()&&validPassword()&&valideImage())
            {
                binding.progressBar.setVisibility(View.VISIBLE);
                // TODO: FireBAse
                mAuth.createUserWithEmailAndPassword(mail, pwd).addOnCompleteListener(new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        if (task.isSuccessful()) {
                            mUser.setUid(mAuth.getCurrentUser().getUid());
                            mUser.setMail(mail);
                            uploadImage();
                        } else {
                            binding.progressBar.setVisibility(View.GONE);
                            Snackbar.make(binding.getRoot(), task.getException().getMessage(), Snackbar.LENGTH_LONG);
                        }

                    }
                });

            }
        }

    }

    private void saveUser() {
        String nom = binding.nomInscri.getText().toString();
        String prenom = binding.prenomInscri.getText().toString();
        mUser.setNom(nom);
        mUser.setPrenom(prenom);
        Log.e("ref", "Im here");
        mRefrence.child(mUser.getUid()).setValue(mUser).addOnCompleteListener(new OnCompleteListener<Void>() {
            @Override
            public void onComplete(@NonNull Task<Void> task) {
                if (task.isSuccessful()) {
                    binding.progressBar.setVisibility(View.GONE);
                    startActivity(new Intent(InscriptionActivity.this, MainActivity.class));
                    finish();
                } else {
                    binding.progressBar.setVisibility(View.GONE);
                    Snackbar.make(binding.getRoot(), task.getException().getMessage(), Snackbar.LENGTH_LONG);
                }
            }
        });
    }

    private void uploadImage() {
        StorageReference child = mStorage.getReference().child(mUser.getUid());
        child.putFile(filePath).continueWithTask(new Continuation<UploadTask.TaskSnapshot, Task<Uri>>() {
            @Override
            public Task<Uri> then(@NonNull Task<UploadTask.TaskSnapshot> task) throws Exception {
                return child.getDownloadUrl();
            }
        }).addOnCompleteListener(new OnCompleteListener<Uri>() {
            @Override
            public void onComplete(@NonNull Task<Uri> task) {
                if (task.isSuccessful()) {
                    mUser.setUrl(task.getResult().toString());
                    Log.e("url", mUser.getUrl());
                    saveUser();
                }

            }
        });
    }


        public void choosePicture(View view) {
            Intent intent = new Intent();
            intent.setType("image/*");
            intent.setAction(Intent.ACTION_OPEN_DOCUMENT);
            startActivityForResult(Intent.createChooser(intent, "Select Picture"), PICK_IMAGE_REQUEST);
        }
        @Override
        protected void onActivityResult(int requestCode, int resultCode, Intent data) {
            super.onActivityResult(requestCode, resultCode, data);
            if (requestCode == PICK_IMAGE_REQUEST && resultCode == RESULT_OK
                    && data != null && data.getData() != null) {
                filePath = data.getData();
                getContentResolver().takePersistableUriPermission(filePath, Intent.FLAG_GRANT_READ_URI_PERMISSION | Intent.FLAG_GRANT_WRITE_URI_PERMISSION);

                try {
                    Bitmap bitmap = MediaStore.Images.Media.getBitmap(getContentResolver(), filePath);
                    binding.addimage.setImageBitmap(bitmap);
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }


    private boolean isEmpty() {
        boolean empty = false ;

        if (TextUtils.isEmpty(binding.mailInscri.getText()))
        {
            empty = true ;
            binding.mailInscri.setError("Champ Vide");
        }
        if ((TextUtils.isEmpty(binding.pwdInscri.getText())))
        {
            empty = true ;
            binding.pwdInscri.setError("Champ Vide");
        }
        if (TextUtils.isEmpty(binding.nomInscri.getText()))
        {
            empty = true ;
            binding.nomInscri.setError("Champ Vide");
        }
        if (TextUtils.isEmpty(binding.prenomInscri.getText()))
        {
            empty = true ;
            binding.prenomInscri.setError("Champ Vide");
        }
        return  empty ;
    }
    private boolean validEmail(){
        boolean valid = true ;
        if (!Patterns.EMAIL_ADDRESS.matcher(binding.mailInscri.getText()).matches())
        {
            valid = false ;
            binding.mailInscri.setError(" entrer un email Valide");
        }

        return valid ;
    }
    private boolean validPassword (){
        boolean valide = true ;
        if (binding.pwdInscri.getText().length() <6)
        {
            valide = false ;
            binding.pwdInscri.setError("minimum 6 caractéres ");
        }
        return  valide ;
    }
    private boolean valideImage (){
        boolean valid = true ;
        if (filePath == null)
        {
            valid = false ;
            Snackbar.make(binding.addimage , "Ajouter une Image ", Snackbar.LENGTH_LONG);
        }
        return  valid ;
    }

    private void initFireBAse(){
        mAuth = FirebaseAuth.getInstance();
        //get la base de donné
        mDatabase = FirebaseDatabase.getInstance();
        // le curseur
        mRefrence = mDatabase.getReference("user");

        mStorage = FirebaseStorage.getInstance();

    }
}

File: app/src/main/java/com/example/firebasestudy/InscriptionActivity.kt
package com.example.firebasestudy

import android.app.Activity
import android.content.Intent
import android.graphics.Bitmap
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import android.text.TextUtils
import android.util.Log
import android.util.Patterns
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.example.firebasestudy.databinding.ActivityInscriptionBinding
import com.example.firebasestudy.model.User
import com.google.android.gms.tasks.Continuation
import com.google.android.gms.tasks.OnCompleteListener
import com.google.android.gms.tasks.Task
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageReference
import java.io.IOException

class InscriptionActivity : AppCompatActivity() {
    private lateinit var binding: ActivityInscriptionBinding
    private var filePath: Uri? = null
    private var mAuth: FirebaseAuth? = null
    private var mDatabase: FirebaseDatabase? = null
    private var mRefrence: DatabaseReference? = null
    private var mStorage: FirebaseStorage? = null
    private val mUser: User? = User()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityInscriptionBinding.inflate(layoutInflater)
        val view: View? = binding.root
        setContentView(view)
        initFireBAse()
        binding.addimage.setOnClickListener { view: View? -> choosePicture(view) }
        binding.inscriBtn.setOnClickListener { save() }
    }

    private fun save() {
        if (!isEmpty()) {
            val mail: String? = binding.mailInscri.text.toString()
            val pwd: String? = binding.pwdInscri.text.toString()
            if (validEmail() && validPassword() && valideImage()) {
                binding.progressBar.visibility = View.VISIBLE
                mAuth?.createUserWithEmailAndPassword(mail!!, pwd!!)?.addOnCompleteListener(OnCompleteListener { task: Task<AuthResult?>? ->
                    if (task!!.isSuccessful) {
                        mUser?.uid=mAuth?.currentUser?.uid
                        mUser?.mail = mail
                        uploadImage()
                    } else {
                        binding.progressBar.visibility = View.GONE
                        Snackbar.make(binding.root, task.exception?.message!!, Snackbar.LENGTH_LONG)
                    }
                })
            }
        }
    }

    private fun uploadImage() {
        val child: StorageReference? = mStorage?.reference?.child(mUser?.uid!!)
        child?.putFile(filePath!!)?.continueWithTask { child.downloadUrl }?.addOnCompleteListener { task: Task<Uri?>? ->
            if (task!!.isSuccessful) {
                mUser?.url=task.result.toString()
                saveUser()
            }
        }
    }

    private fun saveUser() {
        val nom: String? = binding.nomInscri.text.toString()
        mUser?.nom = nom
        mUser?.prenom =binding.prenomInscri.text.toString()
        mRefrence?.child(mUser?.uid!!)?.setValue(mUser)?.addOnCompleteListener { task ->
            if (task.isSuccessful) {
                sendVerificationEmail()
                binding.progressBar.visibility = View.GONE
                startActivity(Intent(this@InscriptionActivity, MainActivity::class.java))
                finish()
            } else {
                binding.progressBar.visibility = View.GONE
                Snackbar.make(binding.root, task.exception?.message!!, Snackbar.LENGTH_LONG)
            }
        }
    }

    private fun sendVerificationEmail() {
        mAuth?.currentUser?.sendEmailVerification()?.addOnCompleteListener(OnCompleteListener { task: Task<Void?>? -> Log.e("mailsend", task.toString()) })
    }

    private fun choosePicture(view: View?) {
        val intent: Intent? = Intent()
        intent?.type = "image/*"
        intent?.action = Intent.ACTION_OPEN_DOCUMENT
        startActivityForResult(Intent.createChooser(intent, "Select Picture"), PICK_IMAGE_REQUEST)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if ((requestCode == PICK_IMAGE_REQUEST) && (resultCode == Activity.RESULT_OK
                        ) && (data != null) && (data.data != null)) {
            filePath = data.data
            contentResolver.takePersistableUriPermission(this.filePath!!, Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION)
            try {
                val bitmap: Bitmap? = MediaStore.Images.Media.getBitmap(contentResolver, filePath)
                binding.addimage.setImageBitmap(bitmap)
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
    }

    private fun isEmpty(): Boolean {
        var empty: Boolean = false
        if (TextUtils.isEmpty(binding.mailInscri.text)) {
            empty = true
            binding.mailInscri.error = "Champ Vide"
        }
        if ((TextUtils.isEmpty(binding.pwdInscri.text))) {
            empty = true
            binding.pwdInscri.error = "Champ Vide"
        }
        if (TextUtils.isEmpty(binding.nomInscri.text)) {
            empty = true
            binding.nomInscri.error = "Champ Vide"
        }
        if (TextUtils.isEmpty(binding.prenomInscri.text)) {
            empty = true
            binding.prenomInscri.error = "Champ Vide"
        }
        return empty
    }

    private fun validEmail(): Boolean {
        var valid: Boolean = true
        if (!Patterns.EMAIL_ADDRESS.matcher(binding.mailInscri.text.toString()).matches()) {
            valid = false
            binding.mailInscri.error = " entrer un email Valide"
        }
        return valid
    }

    private fun validPassword(): Boolean {
        var valide: Boolean = true
        if (binding.pwdInscri.text?.length!! < 6) {
            valide = false
            binding.pwdInscri.error = "minimum 6 caractéres "
        }
        return valide
    }

    private fun valideImage(): Boolean {
        var valid: Boolean = true
        if (filePath == null) {
            valid = false
            Snackbar.make(binding.addimage, "Ajouter une Image ", Snackbar.LENGTH_LONG)
        }
        return valid
    }

    private fun initFireBAse() {
        mAuth = FirebaseAuth.getInstance()
        mDatabase = FirebaseDatabase.getInstance()
        mRefrence = mDatabase?.getReference("user")
        mStorage = FirebaseStorage.getInstance()
    }

    companion object {
        private const val PICK_IMAGE_REQUEST: Int = 1000
    }
}